libminifi/include/agent/agent_docs.h

```diff
@@ -13,53 +13,117 @@
-#include <stdlib.h>
-#include <utils/StringUtils.h>
+#pragma once
 
 #include <map>
 #include <string>
 #include <utility>
+#include <vector>
+
+#include "core/Annotation.h"
+#include "core/Property.h"
+#include "core/Relationship.h"
+#include "utils/Export.h"
+#include "utils/StringUtils.h"
+
+namespace org::apache::nifi::minifi {
+
+enum class ResourceType {
+  Processor, ControllerService, InternalResource, DescriptionOnly
+};
+
+struct ClassDescription {
+  ResourceType type_ = ResourceType::Processor;
+  std::string short_name_{};
+  std::string full_name_{};
+  std::string description_{};
+  std::vector<core::Property> class_properties_{};
+  std::vector<core::Relationship> class_relationships_{};
+  bool dynamic_properties_ = false;
+  bool dynamic_relationships_ = false;
+  std::string inputRequirement_{};
+  bool isSingleThreaded_ = false;
+};
+
+struct Components {
+  std::vector<ClassDescription> processors_;
+  std::vector<ClassDescription> controller_services_;
+  std::vector<ClassDescription> other_components_;
+
+  [[nodiscard]] bool empty() const noexcept {
+    return processors_.empty() && controller_services_.empty() && other_components_.empty();
+  }
+};
+
+namespace detail {
+template<typename Container>
+auto toVector(const Container& container) {
+  return std::vector<typename Container::value_type>(container.begin(), container.end());
+}
+
+template<typename T>
+std::string classNameWithDots() {
+  std::string class_name = core::getClassName<T>();
+  return utils::StringUtils::replaceAll(class_name, "::", ".");
+}
+}  // namespace detail
 
-namespace org {
-namespace apache {
-namespace nifi {
-namespace minifi {
 class AgentDocs {
  private:
-  static std::map<std::string, std::string> &getDescriptions();
+  MINIFIAPI static std::map<std::string, Components> class_mappings_;
 
  public:
-  /**
-   * Updates the internal map with the feature description
-   * @param feature feature ( CS or processor ) whose description is being provided.
-   * @param description provided description.
-   * @return true if update occurred.
-   */
-  static bool putDescription(const std::string &feature, const std::string &description) {
-    return getDescriptions().insert(std::make_pair(feature, description)).second;
+  static const std::map<std::string, Components>& getClassDescriptions() {
+    return class_mappings_;
   }
 
-  /**
-   * Gets the description for the provided feature.
-   * @param feature feature whose description we will provide
-   * @return true if found, false otherwise
-   */
-  static bool getDescription(const std::string &feature, std::string &value) {
-    const std::map<std::string, std::string> &extensions = getDescriptions();
-    auto iff = extensions.find(feature);
-    if (iff != extensions.end()) {
-      value = iff->second;
-      return true;
-    } else {
-      return false;
+  static bool getDescription(const std::string &feature, std::string &value);
+
+  template<typename Class, ResourceType Type>
```

Comment:
The ResourceType could be autodetected IMO. Included some untested detection logic that will probably need some changes, but should at least demonstrate the idea:
```suggestion
  namespace detail {
  template<typename T>
  constexpr bool is_array = false;
  template<typename ValueType, size_t Size>
  constexpr bool is_array<std::array<ValueType, Size>> = true;
  template<typename T, typename ValueType>
  concept array_of = is_array<T> && std::is_same_v<typename T::value_type, ValueType>;
  
  template<typename T>
  concept has_description = requires() {
    { T::Description } -> const char*;
  };
  
  template<typename T>
  concept has_metadata = has_description<T> && requires() {
    { T::properties() } -> array_of<core::Property>;
    { T::SupportsDynamicProperties } -> bool;
  };
  
  template<typename T>
  concept processor = has_description<T> && std::derived_from<T, core::Processor>;
  
  template<typename T>
  concept controller_service = has_description<T> && std::derived_from<T, core::controller::ControllerService>;
  
  struct unknown_resource_type : std::exception {};
  
  template<typename T>
  constexpr ResourceType resource_type_of() {
    if constexpr (processor<T>) {
      return ResourceType::Processor;
    } else if constexpr (controller_service<T>) {
      return ResourceType::ControllerService;
    } else if constexpr (has_metadata<T>) {
      return ResourceType::InternalResource;
    } else if constepxr (has_description<T>) {
      return ResourceType::DescriptionOnly;
    } else {
      throw unknown_resource_type{};
    }
  }
  }  // namespace detail

  template<typename Class, ResourceType Type>
```